File: server/internal/serverstore/grants.go
package serverstore

import (
	"database/sql"
	"strings"
)

// GranteeType distinguishes a direct user grant from a group grant.
type GranteeType string

const (
	GranteeUser  GranteeType = "user"
	GranteeGroup GranteeType = "group"
)

// Grant is one ACL row: a resource granted to a user or a group.
type Grant struct {
	GranteeType GranteeType `json:"grantee_type"`
	Grantee     string      `json:"grantee"`
}

// validGrantee enforces a plain, conflict-free subject name (no separators
// or path-ish characters); group names may carry an optional '@' prefix the
// webadmin sends, stripped before storage.
func validGrantee(g string) (string, bool) {
	g = strings.TrimPrefix(g, "@")
	if g == "" || strings.ContainsAny(g, "/\\\t\n") {
		return "", false
	}
	return g, true
}

// ---- skills ----

// GrantSkill gives a user or group access to a skill (idempotent).
func GrantSkill(db queryer, skillName, grantee string, t GranteeType) error {
	g, ok := validGrantee(grantee)
	if !ok {
		return ErrValidation
	}
	stmt := "INSERT INTO app_grants (kind, app_id, grantee_type, grantee) VALUES ('skill', ?, ?, ?) ON CONFLICT DO NOTHING"
	_, err := db.Exec(stmt, skillName, t, g)
	return err
}

// RevokeSkill removes a grant (idempotent; missing row is not an error).
func RevokeSkill(db queryer, skillName, grantee string, t GranteeType) error {
	g, ok := validGrantee(grantee)
	if !ok {
		return ErrValidation
	}
	_, err := db.Exec("DELETE FROM app_grants WHERE kind = 'skill' AND app_id = ? AND grantee_type = ? AND grantee = ?", skillName, t, g)
	return err
}

// ListSkillGrants returns every grant on a skill.
func ListSkillGrants(db *sql.DB, skillName string) ([]Grant, error) {
	rows, err := db.Query("SELECT grantee_type, grantee FROM app_grants WHERE kind = 'skill' AND app_id = ? ORDER BY grantee_type, grantee", skillName)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []Grant
	for rows.Next() {
		var g Grant
		if err := rows.Scan(&g.GranteeType, &g.Grantee); err != nil {
			return nil, err
		}
		out = append(out, g)
	}
	return out, rows.Err()
}

// AccessibleSkillNames returns the skill names granted to a user directly
// or through any of their groups. Single query, no N+1; a user with no
// grants gets an empty set (strict default: nothing is implicitly visible).
func AccessibleSkillNames(db *sql.DB, username string, groups []string) ([]string, error) {
	var sb strings.Builder
	sb.WriteString("SELECT DISTINCT app_id FROM app_grants WHERE kind = 'skill' AND ((grantee_type = 'user' AND grantee = ?)")
	args := []any{username}
	if len(groups) > 0 {
		// COLLATE NOCASE: LDAP 组名与手输组名大小写差异不得导致授权静默失效
		sb.WriteString(" OR (grantee_type = 'group' AND (")
		for i, g := range groups {
			if i > 0 {
				sb.WriteString(" OR ")
			}
			sb.WriteString(CaseInsensitiveCmp("grantee"))
			args = append(args, g)
		}
		sb.WriteString("))")
	}
	sb.WriteString(")")
	rows, err := db.Query(sb.String(), args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []string
	for rows.Next() {
		var n string
		if err := rows.Scan(&n); err != nil {
			return nil, err
		}
		out = append(out, n)
	}
	return out, rows.Err()
}

// DeleteSkillGrants removes all grants of a skill (resource deletion
// cascades; old grants must never resurrect a re-created resource).
func DeleteSkillGrants(db queryer, skillName string) error {
	_, err := db.Exec("DELETE FROM app_grants WHERE kind = 'skill' AND app_id = ?", skillName)
	return err
}

// ---------------------------------------------------------------------------
// 共享技能/Agent 授权(0036):与 skill_grants 同模型,资源按 name 授权。
// 表名/资源列是编译期常量(grantTable),无用户输入拼接。
// ---------------------------------------------------------------------------

// SharedGrantableTable names one grant table plus the row scope inside it.
//
// P2(迁移 0053):三张旧授权表(skill_grants / shared_skill_grants /
// agent_preset_grants)合并为 app_grants,靠 kind 区分技能与智能体。
// 表名/列名/kind 都是编译期常量,不接受用户输入拼接。
type SharedGrantableTable struct {
	Table string // 统一为 "app_grants"
	Col   string // 资源名列("app_id")
	Kind  string // 行作用域:"skill" | "agent"(空 = 不加 kind 条件)
}

// scope 返回附加的 kind 条件与参数(kind 为空时不加条件)。
func (t SharedGrantableTable) scope() (string, []any) {
	if t.Kind == "" {
		return "", nil
	}
	return " AND kind = ?", []any{t.Kind}
}

var (
	// SharedSkillGrantTable: 技能授权(市场与组织共用一个命名空间)。
	SharedSkillGrantTable = SharedGrantableTable{Table: "app_grants", Col: "app_id", Kind: "skill"}
	// SharedPresetGrantTable: 智能体授权。
	SharedPresetGrantTable = SharedGrantableTable{Table: "app_grants", Col: "app_id", Kind: "agent"}
)

// GrantSharedResource gives a user or group access to a shared skill/preset
// (idempotent). The table must be one of the compile-time constants above.
func GrantSharedResource(db queryer, table SharedGrantableTable, resourceName, grantee string, t GranteeType) error {
	g, ok := validGrantee(grantee)
	if !ok {
		return ErrValidation
	}
	stmt := "INSERT INTO " + table.Table + " (kind, " + table.Col + ", grantee_type, grantee) VALUES (?, ?, ?, ?) ON CONFLICT DO NOTHING"
	_, err := db.Exec(stmt, table.Kind, resourceName, t, g)
	return err
}

// RevokeSharedResource removes a grant (idempotent).
func RevokeSharedResource(db queryer, table SharedGrantableTable, resourceName, grantee string, t GranteeType) error {
	g, ok := validGrantee(grantee)
	if !ok {
		return ErrValidation
	}
	cond, args := table.scope()
	args = append([]any{resourceName, t, g}, args...)
	_, err := db.Exec("DELETE FROM "+table.Table+" WHERE "+table.Col+" = ? AND grantee_type = ? AND grantee = ?"+cond, args...)
	return err
}

// ListSharedResourceGrants returns every grant on a shared skill/preset.
func ListSharedResourceGrants(db *sql.DB, table SharedGrantableTable, resourceName string) ([]Grant, error) {
	cond, extra := table.scope()
	rows, err := db.Query("SELECT grantee_type, grantee FROM "+table.Table+" WHERE "+table.Col+" = ?"+cond+" ORDER BY grantee_type, grantee",
		append([]any{resourceName}, extra...)...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []Grant
	for rows.Next() {
		var g Grant
		if err := rows.Scan(&g.GranteeType, &g.Grantee); err != nil {
			return nil, err
		}
		out = append(out, g)
	}
	return out, rows.Err()
}

// DeleteSharedResourceGrants removes all grants of a resource (delete
// cascades; old grants must never resurrect a re-created resource).
func DeleteSharedResourceGrants(db queryer, table SharedGrantableTable, resourceName string) error {
	cond, extra := table.scope()
	_, err := db.Exec("DELETE FROM "+table.Table+" WHERE "+table.Col+" = ?"+cond, append([]any{resourceName}, extra...)...)
	return err
}

// ReplaceSharedGroups sets the full group-grant set of a shared resource in
// one transaction (existing group grants dropped; user grants untouched).
// Every name must reference an existing department.
func ReplaceSharedGroups(db *sql.DB, table SharedGrantableTable, resourceName string, groups []string) error {
	cond, extra := table.scope()
	return replaceGroupGrants(db,
		"DELETE FROM "+table.Table+" WHERE "+table.Col+" = ? AND grantee_type = 'group'"+cond,
		append([]any{resourceName}, extra...),
		func(tx *sql.Tx, name string) error {
			_, err := tx.Exec("INSERT INTO "+table.Table+" (kind, "+table.Col+", grantee_type, grantee) VALUES (?, ?, 'group', ?)",
				table.Kind, resourceName, name)
			return err
		},
		groups)
}

// AccessibleSharedResourceNames returns the shared skill/preset names granted
// to a user directly or through any of their groups (strict default).
func AccessibleSharedResourceNames(db *sql.DB, table SharedGrantableTable, username string, groups []string) ([]string, error) {
	var sb strings.Builder
	sb.WriteString("SELECT DISTINCT " + table.Col + " FROM " + table.Table + " WHERE ")
	args := []any{}
	if cond, extra := table.scope(); cond != "" {
		sb.WriteString("kind = ? AND ")
		args = append(args, extra...)
	}
	sb.WriteString("((grantee_type = 'user' AND grantee = ?)")
	args = append(args, username)
	if len(groups) > 0 {
		sb.WriteString(" OR (grantee_type = 'group' AND (")
		for i, g := range groups {
			if i > 0 {
				sb.WriteString(" OR ")
			}
			sb.WriteString(CaseInsensitiveCmp("grantee"))
			args = append(args, g)
		}
		sb.WriteString("))")
	}
	sb.WriteString(")")
	rows, err := db.Query(sb.String(), args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []string
	for rows.Next() {
		var n string
		if err := rows.Scan(&n); err != nil {
			return nil, err
		}
		out = append(out, n)
	}
	return out, rows.Err()
}

// ---- 整组替换(多部门批量授权,原子) ----

// replaceGroupGrants replaces all group grants of a resource in one
// transaction: existing group grants are dropped, the given department
// names become the full group-grant set (user grants untouched).
// Every name must reference an existing department (typos fail fast).
func replaceGroupGrants(db *sql.DB, deleteSQL string, deleteArgs []any, insert func(tx *sql.Tx, name string) error, groups []string) error {
	tx, err := db.Begin()
	if err != nil {
		return err
	}
	defer tx.Rollback()
	// 部门存在性校验放事务内(审计 A5-L9):与写入同事务,消除
	// 「校验通过后、写事务前部门被删」的 TOCTOU 窗口,不留孤儿授权行。
	// 只接受已存在部门,防拼错隐式建组;重复/空名直接拒绝。
	seen := map[string]bool{}
	for _, g := range groups {
		if g == "" || seen[g] {
			return ErrValidation
		}
		seen[g] = true
		var n int
		if err := tx.QueryRow("SELECT COUNT(*) FROM groups WHERE "+CaseInsensitiveCmp("name"), g).Scan(&n); err != nil {
			return err
		}
		if n == 0 {
			return ErrNotFound
		}
	}
	if _, err := tx.Exec(deleteSQL, deleteArgs...); err != nil {
		return err
	}
	for _, g := range groups {
		if err := insert(tx, g); err != nil {
			return err
		}
	}
	return tx.Commit()
}

// ReplaceSkillGroupGrants sets the full department-grant set of a skill.
func ReplaceSkillGroupGrants(db *sql.DB, skillName string, groups []string) error {
	return replaceGroupGrants(db,
		"DELETE FROM app_grants WHERE kind = 'skill' AND app_id = ? AND grantee_type = 'group'", []any{skillName},
		func(tx *sql.Tx, name string) error {
			_, err := tx.Exec("INSERT INTO app_grants (kind, app_id, grantee_type, grantee) VALUES ('skill', ?, 'group', ?)", skillName, name)
			return err
		},
		groups)
}

File: rust/crates/dsh-store/src/grants.rs
//! 授权域（Go `serverstore/grants.go` 等价）——app_grants ACL。

use crate::errors::{map_db_error, StoreError};
use sqlx::Row;

/// GranteeType 区分用户直授与组授。
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum GranteeType {
    User,
    Group,
}

impl GranteeType {
    pub fn as_str(&self) -> &'static str {
        match self {
            GranteeType::User => "user",
            GranteeType::Group => "group",
        }
    }
}

/// Grant 一条 ACL 行：资源授权给用户或组。
#[derive(Debug, Clone, Default)]
pub struct Grant {
    pub grantee_type: String,
    pub grantee: String,
}

/// SharedGrantableTable 命名授权行作用域（编译期常量，无用户输入拼接）。
#[derive(Debug, Clone, Copy)]
pub struct SharedGrantableTable {
    pub kind: &'static str, // "skill" | "agent" | ""
}

pub const SHARED_SKILL_GRANT_TABLE: SharedGrantableTable = SharedGrantableTable { kind: "skill" };
pub const SHARED_PRESET_GRANT_TABLE: SharedGrantableTable = SharedGrantableTable { kind: "agent" };

/// valid_grantee 强制普通、无冲突主题名（无分隔符/路径字符；组名可带 @ 前缀剥离）。
fn valid_grantee(g: &str) -> Result<String, StoreError> {
    let g = g.strip_prefix('@').unwrap_or(g);
    if g.is_empty() || g.contains(['/', '\\', '\t', '\n']) {
        return Err(StoreError::Validation);
    }
    Ok(g.to_string())
}

/// GrantSkill 授予用户/组技能访问（幂等）。
pub async fn grant_skill(
    pool: &sqlx::PgPool,
    skill_name: &str,
    grantee: &str,
    t: GranteeType,
) -> Result<(), StoreError> {
    let g = valid_grantee(grantee)?;
    sqlx::query(
        "INSERT INTO app_grants (kind, app_id, grantee_type, grantee) VALUES ('skill', $1, $2, $3) ON CONFLICT DO NOTHING",
    )
    .bind(skill_name)
    .bind(t.as_str())
    .bind(&g)
    .execute(pool)
    .await
    .map_err(map_db_error)?;
    Ok(())
}

/// RevokeSkill 移除授权（幂等；缺失行非错误）。
pub async fn revoke_skill(
    pool: &sqlx::PgPool,
    skill_name: &str,
    grantee: &str,
    t: GranteeType,
) -> Result<(), StoreError> {
    let g = valid_grantee(grantee)?;
    sqlx::query("DELETE FROM app_grants WHERE kind = 'skill' AND app_id = $1 AND grantee_type = $2 AND grantee = $3")
        .bind(skill_name)
        .bind(t.as_str())
        .bind(&g)
        .execute(pool)
        .await
        .map_err(map_db_error)?;
    Ok(())
}

/// ListSkillGrants 返回技能的全部授权。
pub async fn list_skill_grants(pool: &sqlx::PgPool, skill_name: &str) -> Result<Vec<Grant>, StoreError> {
    let rows = sqlx::query(
        "SELECT grantee_type, grantee FROM app_grants WHERE kind = 'skill' AND app_id = $1 ORDER BY grantee_type, grantee",
    )
    .bind(skill_name)
    .fetch_all(pool)
    .await
    .map_err(map_db_error)?;
    Ok(rows.into_iter().map(scan_grant).collect())
}

fn scan_grant(r: sqlx::postgres::PgRow) -> Grant {
    Grant {
        grantee_type: r.get("grantee_type"),
        grantee: r.get("grantee"),
    }
}

/// AccessibleSkillNames 用户直授或经组授得到的技能名（严格默认：无授权不可见）。
pub async fn accessible_skill_names(
    pool: &sqlx::PgPool,
    username: &str,
    groups: &[String],
) -> Result<Vec<String>, StoreError> {
    accessible_app_ids(pool, "skill", username, groups).await
}

/// accessible_app_ids 通用实现。
pub async fn accessible_app_ids(
    pool: &sqlx::PgPool,
    kind: &str,
    username: &str,
    groups: &[String],
) -> Result<Vec<String>, StoreError> {
    let mut sql = String::from(
        "SELECT DISTINCT app_id FROM app_grants WHERE kind = $1 AND (grantee_type = 'user' AND grantee = $2",
    );
    let mut idx = 3usize;
    if !groups.is_empty() {
        sql.push_str(" OR (grantee_type = 'group' AND (");
        for (i, _) in groups.iter().enumerate() {
            if i > 0 {
                sql.push_str(" OR ");
            }
            sql.push_str(&format!("LOWER(grantee) = LOWER(${idx})"));
            idx += 1;
        }
        sql.push_str("))");
    }
    sql.push_str(")");
    let mut q = sqlx::query(&sql).bind(kind).bind(username);
    for g in groups {
        q = q.bind(g);
    }
    let rows = q.fetch_all(pool).await.map_err(map_db_error)?;
    Ok(rows.into_iter().map(|r| r.get::<String, _>("app_id")).collect())
}

/// DeleteSkillGrants 清除技能全部授权（资源删除级联）。
pub async fn delete_skill_grants(pool: &sqlx::PgPool, skill_name: &str) -> Result<(), StoreError> {
    sqlx::query("DELETE FROM app_grants WHERE kind = 'skill' AND app_id = $1")
        .bind(skill_name)
        .execute(pool)
        .await
        .map_err(map_db_error)?;
    Ok(())
}

/// GrantSharedResource 授予共享技能/智能体（幂等）。
pub async fn grant_shared_resource(
    pool: &sqlx::PgPool,
    table: SharedGrantableTable,
    resource_name: &str,
    grantee: &str,
    t: GranteeType,
) -> Result<(), StoreError> {
    let g = valid_grantee(grantee)?;
    sqlx::query("INSERT INTO app_grants (kind, app_id, grantee_type, grantee) VALUES ($1, $2, $3, $4) ON CONFLICT DO NOTHING")
        .bind(table.kind)
        .bind(resource_name)
        .bind(t.as_str())
        .bind(&g)
        .execute(pool)
        .await
        .map_err(map_db_error)?;
    Ok(())
}

/// RevokeSharedResource 移除共享资源授权（幂等）。
pub async fn revoke_shared_resource(
    pool: &sqlx::PgPool,
    table: SharedGrantableTable,
    resource_name: &str,
    grantee: &str,
    t: GranteeType,
) -> Result<(), StoreError> {
    let g = valid_grantee(grantee)?;
    sqlx::query("DELETE FROM app_grants WHERE app_id = $1 AND grantee_type = $2 AND grantee = $3 AND kind = $4")
        .bind(resource_name)
        .bind(t.as_str())
        .bind(&g)
        .bind(table.kind)
        .execute(pool)
        .await
        .map_err(map_db_error)?;
    Ok(())
}

/// ListSharedResourceGrants 返回共享技能/智能体的全部授权。
pub async fn list_shared_resource_grants(
    pool: &sqlx::PgPool,
    table: SharedGrantableTable,
    resource_name: &str,
) -> Result<Vec<Grant>, StoreError> {
    let rows = sqlx::query(
        "SELECT grantee_type, grantee FROM app_grants WHERE kind = $1 AND app_id = $2 ORDER BY grantee_type, grantee",
    )
    .bind(table.kind)
    .bind(resource_name)
    .fetch_all(pool)
    .await
    .map_err(map_db_error)?;
    Ok(rows.into_iter().map(scan_grant).collect())
}

/// DeleteSharedResourceGrants 清除共享资源全部授权。
pub async fn delete_shared_resource_grants(
    pool: &sqlx::PgPool,
    table: SharedGrantableTable,
    resource_name: &str,
) -> Result<(), StoreError> {
    sqlx::query("DELETE FROM app_grants WHERE kind = $1 AND app_id = $2")
        .bind(table.kind)
        .bind(resource_name)
        .execute(pool)
        .await
        .map_err(map_db_error)?;
    Ok(())
}

/// AccessibleSharedResourceNames 用户直授或经组授得到的共享资源名。
pub async fn accessible_shared_resource_names(
    pool: &sqlx::PgPool,
    table: SharedGrantableTable,
    username: &str,
    groups: &[String],
) -> Result<Vec<String>, StoreError> {
    let mut sql = String::from(
        "SELECT DISTINCT app_id FROM app_grants WHERE kind = $1 AND (grantee_type = 'user' AND grantee = $2",
    );
    let mut idx = 3usize;
    if !groups.is_empty() {
        sql.push_str(" OR (grantee_type = 'group' AND (");
        for (i, _) in groups.iter().enumerate() {
            if i > 0 {
                sql.push_str(" OR ");
            }
            sql.push_str(&format!("LOWER(grantee) = LOWER(${idx})"));
            idx += 1;
        }
        sql.push_str("))");
    }
    sql.push_str(")");
    let mut q = sqlx::query(&sql).bind(table.kind).bind(username);
    for g in groups {
        q = q.bind(g);
    }
    let rows = q.fetch_all(pool).await.map_err(map_db_error)?;
    Ok(rows.into_iter().map(|r| r.get("app_id")).collect())
}

/// ReplaceSharedGroups 事务内整体替换共享资源的组授权集（用户授权不动）。
pub async fn replace_shared_groups(
    pool: &sqlx::PgPool,
    table: SharedGrantableTable,
    resource_name: &str,
    groups: &[String],
) -> Result<(), StoreError> {
    let mut tx = pool.begin().await.map_err(map_db_error)?;
    // 部门存在性校验放事务内（TOCTOU）
    let mut seen = std::collections::HashSet::new();
    for g in groups {
        if g.is_empty() || !seen.insert(g.to_string()) {
            return Err(StoreError::Validation);
        }
        let n: i64 = sqlx::query_scalar("SELECT COUNT(*) FROM groups WHERE LOWER(name) = LOWER($1)")
            .bind(g)
            .fetch_one(&mut *tx)
            .await
            .map_err(map_db_error)?;
        if n == 0 {
            return Err(StoreError::NotFound);
        }
    }
    sqlx::query("DELETE FROM app_grants WHERE kind = $1 AND app_id = $2 AND grantee_type = 'group'")
        .bind(table.kind)
        .bind(resource_name)
        .execute(&mut *tx)
        .await
        .map_err(map_db_error)?;
    for g in groups {
        sqlx::query("INSERT INTO app_grants (kind, app_id, grantee_type, grantee) VALUES ($1, $2, 'group', $3)")
            .bind(table.kind)
            .bind(resource_name)
            .bind(g)
            .execute(&mut *tx)
            .await
            .map_err(map_db_error)?;
    }
    tx.commit().await.map_err(map_db_error)?;
    Ok(())
}

/// ReplaceSkillGroupGrants 整体替换技能的部门授权集。
pub async fn replace_skill_group_grants(
    pool: &sqlx::PgPool,
    skill_name: &str,
    groups: &[String],
) -> Result<(), StoreError> {
    replace_shared_groups(pool, SHARED_SKILL_GRANT_TABLE, skill_name, groups).await
}

/// DeleteAppGrants 清空某 App 全部授权（资源删除级联）。
pub async fn delete_app_grants(pool: &sqlx::PgPool, kind: &str, app_id: &str) -> Result<(), StoreError> {
    sqlx::query("DELETE FROM app_grants WHERE kind = $1 AND app_id = $2")
        .bind(kind)
        .bind(app_id)
        .execute(pool)
        .await
        .map_err(map_db_error)?;
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::testutil::new_test_db;

    #[tokio::test]
    async fn grants_skill() {
        let pool = new_test_db().await;
        // 建组
        let gid = crate::departments::create_department(&pool, "研发组", 0, 0, "").await.unwrap();
        grant_skill(&pool, "my-skill", "alice", GranteeType::User).await.unwrap();
        grant_skill(&pool, "my-skill", "研发组", GranteeType::Group).await.unwrap();
        let grants = list_skill_grants(&pool, "my-skill").await.unwrap();
        assert_eq!(grants.len(), 2);
        // accessible: 用户直授
        let names = accessible_skill_names(&pool, "alice", &[]).await.unwrap();
        assert!(names.contains(&"my-skill".to_string()));
        // 组授
        let names = accessible_skill_names(&pool, "bob", &["研发组".to_string()]).await.unwrap();
        assert!(names.contains(&"my-skill".to_string()));
        // 非法 grantee
        assert_eq!(
            grant_skill(&pool, "s", "a/b", GranteeType::User).await.unwrap_err(),
            StoreError::Validation
        );
        // 整组替换
        replace_skill_group_grants(&pool, "my-skill", &["研发组".to_string()]).await.unwrap();
        let grants = list_skill_grants(&pool, "my-skill").await.unwrap();
        assert_eq!(grants.len(), 2); // user 授权保留 + 组授权
    }
}
